ejb/WFLY-13354-galleon-for-ejb.adoc

```diff
@@ -0,0 +1,110 @@
+* https://issues.redhat.com/browse/WFLY-13433[Improve capability support in EJB3 subsystem]
+* https://issues.redhat.com/browse/WFLY-13356[Add a Galleon layer for Webservices]
+
+=== Dev Contacts
+
+* mailto:{email}[{author}]
+
+=== QE Contacts
+
+* mailto:fburzigo@redhat.com[Fabio Burzigotti]
+
+=== Testing By
+
+* [X] Engineering
+
+* [ ] QE
+
+=== Affected Projects or Components
+
+WildFly
+
+=== Other Interested Projects
+
+== Requirements
+
+These analysis and requirements are based on https://github.com/bstansberry/wildfly/commits/ejb-layers2[Brian's proposal].
+
+If a user is trying to slim their server I think of four key things they want to do in descending order of importance:
+
+1. Eliminate unnecessary open ports.
+
+2. Eliminate unnecessary exposure.
+
+3. Eliminate jars, both to save filesystem footprint and to reduce any theoretical attack surface.
+
+4. Reduce configuration clutter and unnecessary services.
+
+Based on the above factors, the following layers are to be supported:
+
+1. ejb3-lite. This is like what we provide in standalone.xml, but without the MDB instance pool or the remote connector resource.
+(A webservices layer would depend on this.)
+
+2. ejb3. Builds (i.e. depends) on ejb3-lite and adds the discovery subsystem, the MDB pool and the remote connector resource.
+This looks what we provide in standalone-full.xml, except no IIOP integration.
+
+3. ejb3-iiop. Builds (i.e. depends) on *ejb3-lite* and adds the iiop subsystem plus the integration resource in the EJB3 subsystem.
+This *could* depend on *ejb3* instead, but that means EJB is unnecessarily exposed over the HTTP interface and a lot of
+dependencies are brought in for messaging integration.
+
+A user could provision ejb3 + ejb3-iiop and get the equivalent to what's in standalone-full.xml.
+
+There'd be a couple ancillary layers as well:
+
+1. ejb-local-cache. This provides the infinispan subsystem resources related to local ejb caching
+(i.e. the ones in standalone.xml and standalone-full.xml.) The ejb3-lite layer *optionally* depends on this.
+It's an optional dependency because the user when provisioning can *exclude* this layer and instead provision...
+
+2. ejb-dist-cache. This provides the infinispan subsystem resources related to distributed ejb caching
+(i.e. the ones in standalone0ha.xml and standalone-full-ha.xml.)
+
+This pattern of "exclude an optional local caching layer and add a distributed variant" is how web session
+and jpa caching are handled in the layers for those features.
+
+TBH I don't see big FS footprint improvements coming out of these permutations. Big EJB dependencies include (in no particular order.)
+
+* IIOP. But the TM uses IIOP libs so they will be there regardless.
+* Remoting. It would be nice to be able to eliminate the need for remoting, e.g. for ejb3-lite if the user app isn't a remote ejb client. But that doesn't seem trivial and if a server is manageable remoting will be there anyway to support the CLI.
```

Comment:
FY the remoting layer is going to be needing some work, I am seeing it is defaulting to a dependency on legacy security realms without setting up the correct dependencies but also needs cleaning up to use elytron.